### DISCUSSION TITLE
Windows Native Command to PowerShell Equivalent Discussion

### DISCUSSION THREAD

**Discussion by StevenBucher98** [PowerShell/PowerShell#19674] [Enhancements]
What kind of other native commands in Windows that have an equivalent PowerShell cmdlet? Feel free to add any you know of! We can look at making a feedback provider around these cmdlets to inform the users

**Comment by StevenBucher98** [comment]
One someone mentioned is `netstat` and is PowerShell equivalent  `Get-NetTcpConnection`

**Comment by StartAutomating** [comment]
The high level real problem here is this:  _How does a command say what native commands are equivalents?_

I've found that there's a lot you can do with small "attribute handshakes".

Essentially, you can not just look for a command, but any command with a specific attribute, and you can use the presence of the attribute to indicate that the command should work a particular way.

[Reflection.AssemblyMetdata] is a useful and open-ended attribute that allows for a single key-value pair.

So, for instance, I could add:

~~~PowerShell
[Reflection.AssemblyMetadata("Wraps","git")]
~~~

To the top of Use-Git, to indicate it is a Wrapper for "git".

Then I could `Get-Command` and find all commands that have this attribute and pick out each value, thus returning something like:

|Command | Wraps|
|-|-|
|Use-Git|git|

If, as a community, we picked an attribute to standardize around, this should be _very_ easy to add to existing commands, and integrate into the engine.

Then the PowerShell engine could do something like:

~~~PowerShell
Get-Command -Wrapper  
~~~

To get back all of the commands that wrap a native command.


One final note for the moment:

While I _really_ think this approach is the right one for exposing this more sanely / broadly that adding something that would only work in very recent PowerShell, I'm not sold on the name of the metadata handshake.  Very open to suggestions here.

**Comment by SteveL-MSFT** [comment]
I think this Feedback Provider should be a self-contained module published to PowerShellGallery with its own repo that the community can contribute to for the mapping.  With this model, it could point users to Crescendo or other modules that wrap native commands and not just the built-in Windows cmdlets.

**Reply by StartAutomating** [reply]
So a module of _just_ the mappings?

I could get behind this, and try to make something pretty easy to edit (long live workflows).

**Comment by Ayanmullick** [comment]
I look for a cmdlet in the PowerShell Module browser first. Eg. One can search for `dns` and [find `Resolve-DnsName` easily][1], @compwiz32.


[Searching `NetTcpConnection`][2] does get one to `Get-NetTCPConnection`. However, if [the online help][3] included that `Get-NetTcpConnection` was the equivalent of `netstat` and made that searchable on the browser it could enable discovery of such cmdlets.

Otherwise, [if one had Chocolatey registered][4], one could get the below list when searching for `netstat` to explore further.

```pwsh
Find-Module *netstat*|
        Select-Object Version,Name,Author,CompanyName,Copyright,PublishedDate,PackageManagementProvider,ProjectUri,RepositorySourceLocation,Description|
            Format-Table
```

<details> <summary>Output</summary>

![image](https://github.com/PowerShell/PowerShell/assets/17693249/e29d2512-ba26-452f-9405-cc664af777f3)

</details>

Ideally, one should be able to find `Get-NetTCPConnection` searching for all TCP related cmdlets in network related modules like below.

`((Find-Module *net*).AdditionalMetadata.Functions -split ' ').Where{$_ -like '*tcp*'}`

However, this doesn't work since the [NetTCPIP][5] module [**isn't published in the gallery**][6], @StevenBucher98 . However, running `Get-Command -Verb Get -Noun *TCP*` locally gets you there.

[1]: https://learn.microsoft.com/en-us/powershell/module/?term=dns#:~:text=functioning%20DNS%20server.-,Resolve%2DDnsName,-Command
[2]: https://learn.microsoft.com/en-us/powershell/module/?term=NetTcpConnection
[3]: https://learn.microsoft.com/en-us/powershell/module/nettcpip/get-nettcpconnection
[4]: https://serverfault.com/a/633648/191518
[5]: https://learn.microsoft.com/en-us/powershell/module/nettcpip/?view=windowsserver2022-ps
[6]: https://www.powershellgallery.com/packages?q=NetTCPIP
